### PR TITLE
Wings: Fix stoplogic

### LINKED ISSUE #40: Wings does not see the stopping state causing servers to be marked as crashed
## Description

For some eggs, when you hit stop, the server is marked as "crashed" even though it stops correctly.

## Steps to Reproduce

Add the following logging to the `server/server.go` file to observe the state changes around the crash logic:

```go
// Log the values being compared
fmt.Printf("Not in crash\n")
fmt.Printf("prevState: %v\n", prevState)
fmt.Printf("Current Environment State: %v\n", s.Environment.State())

if (prevState == environment.ProcessStartingState || prevState == environment.ProcessRunningState) && s.Environment.State() == environment.ProcessOfflineState {
    // Log the values being compared
    fmt.Printf("In Crash\n")
    fmt.Printf("prevState: %v\n", prevState)
    fmt.Printf("Current Environment State: %v\n", s.Environment.State())
}
```

## What You Will See

For example, when a simple Node.js server stops, you will see this:

afbeelding

The "wings" do not see the server going through the stopping state, even though the panel console shows that it went through the stopping state (as per console messages).
When It Happens

If you remove the following commits (just for testing):

afbeelding

The panel correctly sees the server going through the stopping state:

afbeelding
## What I Think Is Happening

~~It seems that because the requests occur so quickly one after another, only the last one is recorded, causing the stopping state to be skipped.~~

It seems the crash detection logic is running after the 2 state changes were made
afbeelding

### PR DESCRIPTION
**Pull request by QuintenQVD0**
# Changes

- Instead of  ContainerKill we use ContainerStop
- This takes 3 args, the ctx, the container id and the container.StopOptions what takes a timeout (as an int ) and a signal as a string
- NEEDS: https://github.com/pelican-dev/panel/pull/407
- closes #40 

The signal must be a string so "SIGTERM" but we used to pass to syscall.SIGTERM and trim some parts but that does not work.
so in the function, all incoming signals are converted to signals in string format. They also get matched with their corresponding timeout. What means for any signal the default timeout before it gets killed is 10 seconds. So:
1. The container gets the signal 
2. it sends the signal to the container
3. if it exitis in those 10 seconds then it will just continue
4. if not after 10 seconds, SIGKILL is sent

The timeout is 0 for os.kill as that is what the kill button sends and then it should stop immediately .

this is combined with my panel PR so `^C` send SIGTERM.

This also makes wine eggs not need anymore for `^^C` but wine can not wait it will just direct exit (how it also does it with `^^C`)

you can also now do as stop `^SIGINT`, `^SIGTERM`, `^SIGABRT` what will be converted in to a signal.

This would stop the need for tini but it still can be usefull.

### mordhau:

afbeelding
Tested with the panel patch with `^C` and `^SIGTERM`

### CODE REVIEW

File: environment/docker/power.go [hunk 1]
```diff
@@ -290,10 +290,47 @@ func (e *Environment) Terminate(ctx context.Context, signal os.Signal) error {
...
+	case syscall.SIGABRT:
+		sig = "SIGABRT"
+		noWaitTimeout = 10
+	case syscall.SIGINT:
+		sig = "SIGINT"
+		noWaitTimeout = 10
+	case syscall.SIGTERM:
+		sig = "SIGTERM"
+		noWaitTimeout = 10
+	case os.Kill:
```
Comment (RMartinOscar) [inline environment/docker/power.go:324]:
```suggestion
```

File: environment/docker/power.go [hunk 2]
```diff
@@ -290,10 +290,47 @@ func (e *Environment) Terminate(ctx context.Context, signal os.Signal) error {
...
+		sig = "SIGKILL"
```
Comment (RMartinOscar) [inline environment/docker/power.go:325]:
```suggestion
```

File: environment/docker/power.go [hunk 3]
```diff
@@ -290,10 +290,47 @@ func (e *Environment) Terminate(ctx context.Context, signal os.Signal) error {
...
+		noWaitTimeout = 0
```
Comment (RMartinOscar) [inline environment/docker/power.go:326]:
```suggestion
```

File: environment/docker/power.go (hunk 1 above)
Comment (RMartinOscar) [inline environment/docker/power.go:324]:
```suggestion

```

File: environment/docker/power.go (hunk 2 above)
Comment (RMartinOscar) [inline environment/docker/power.go:325]:
```suggestion

```

File: environment/docker/power.go (hunk 3 above)
Comment (RMartinOscar) [inline environment/docker/power.go:326]:
```suggestion
```

**Review by RMartinOscar — CHANGES_REQUESTED** [review]
```suggestion
	case os.Kill:
		sig = "SIGKILL"
		noWaitTimeout = 0
```
Those are useles because of the default case

**Comment by QuintenQVD0** [comment]
the docker default is SIGTERM and a timeout of 10seconds? and sigkill must happen imiditaly ?

**Comment by RMartinOscar** [comment]
There's no use to match case for os.Kill when the consequence is the same as the default case

**Comment by QuintenQVD0** [comment]
> There's no use to match case for os.Kill when the consequence is the same as the default case

I changed it.

**Comment by QuintenQVD0** [comment]
Converted to a draft as their is something wrong with the wings or panel logic itself

**Comment by QuintenQVD0** [comment]
IF you revert the this commit: https://github.com/pelican-dev/wings/commit/7bf9f6a9d0f8061bcdbcceeb34372d0f32ad25e6

then stop works but kill is broken

**Comment by RMartinOscar** [comment]
[This](https://github.com/pterodactyl/wings/pull/192) may resolve it

**Comment by QuintenQVD0** [comment]
> [This](https://github.com/pterodactyl/wings/pull/192) may resolve it

I worked with him and no. Their is something weong with our panel reporting container statuses. I can't realy point to what exactly but this pr does the same just a diffrend way.

**Comment by notAreYouScared** [comment]
@QuintenQVD0 Any update with this?

**Comment by QuintenQVD0** [comment]
> @QuintenQVD0 Any update with this?
> 

I have been working a lot latly so my time is limited. The logic is right but there something else preventing this to work with the state of the container.  (most servers in pelican even on properly stop mark it as crashed) that it is skipping the stopping state and so it thinks it has crashed. So until that is fixt this should not be merged as then a server can not be stoped anymore.

 Something is or went wrong with eather a panel update or the pr of Kubi to signal the panel of a state changes.

**Comment by QuintenQVD0** [comment]
Moral of this PR.

Fix the broken stop logic for signals (if https://github.com/pelican-dev/panel/pull/407 is also merged ofc)
Fix the crash detection issue because of a pending api call to the panel that had to finish before the crash detection could running causing it not to see the server go thru the stopping state.

File: environment/docker/power.go [hunk 4]
```diff
@@ -149,16 +148,18 @@ func (e *Environment) Stop(ctx context.Context) error {
 			log.WithField("container_id", e.Id).Warn("no stop configuration detected for environment, using termination procedure")
 		}
 
-		signal := os.Kill
+		signal := "SIGKILL"
```
Comment (parkervcp) [inline environment/docker/power.go:151]: This could be moved into the switch as the default case

**Review by parkervcp — CHANGES_REQUESTED** [review]

**Review by parkervcp — APPROVED** [review]
LGTM

**Pull request merged**